outlines/text/generate/cfg.py

```diff
@@ -0,0 +1,300 @@
+import math
+from typing import TYPE_CHECKING, List, Optional, Union
+
+import torch
+from lark import Lark, UnexpectedCharacters, UnexpectedToken
+
+from outlines.text.generate.continuation import Continuation
+from outlines.text.generate.regex import regex
+
+if TYPE_CHECKING:
+    from outlines.text.generate.regex import Regex
+    from outlines.text.generate.sample import Sampler
+
+
+class CFG(Continuation):
+    """Represents a cfg-based generation model.
+
+    `CFG` instances are constrained generation models that only generate
+    sequences matching a given context-free grammar.
+
+    >>> import outlines.text as text
+    >>> grammar = '''
+    ...     start: palindrome
+    ...     palindrome: letter
+    ...         | "a" palindrome "a"
+    ...         | "b" palindrome "b"
+    ...         | "c" palindrome "c"
+    ...     letter: "a" | "b" | "c"
+    ... '''
+    >>> generator = text.generate.cfg(model, grammar)
+
+    Sequences can then be generated from a prompt as follows:
+
+    >>> sequence = generator("Return a palindrome using only the characters 'a', 'b', and 'c': ")
+
+    .. note:
+        Reuse instances of these guided generators (e.g. `generator` from the
+        above example) whenever possible, because constructing them has more
+        overhead than generating token sequences from them.
+
+    """
+
+    def __init__(
+        self,
+        model,
+        cfg_string: str,
+        max_tokens: Optional[int] = None,
+        *,
+        sampler: Optional["Sampler"] = None,
+        stop: Union[str, List[str]] = [],
+        allow_empty_tokens: bool = True,
+    ):
+        """
+
+        Parameters
+        ----------
+        model
+            The instance of the model used to generate next-token probabilities.
+        cfg_string
+            The cfg with which the token sampling process is guided/constrained.
+        max_tokens
+            The maximum number of tokens to be sampled.
+        sampler
+            The function used to draw samples.  Defaults to
+            `outlines.text.generate.sample.multinomial`.  See
+            `outlines.text.generate.sample.Sampler` for the expected form of
+            such functions.
+        stop
+            Optional stopping string(s).
+        allow_empty_tokens
+            Allow sampling of tokens corresponding to empty strings.
+
+        """
+        super().__init__(model, max_tokens, sampler, stop)
+        self.allow_empty_tokens = allow_empty_tokens
+
+        self.parser = Lark(
```

Comment:
Note that we have a customized `lark`-based parser implementation in `outlines.text.parsing.PartialLark` that avoids all the unnecessary scanner backtracking implicit to a non-incremental parser approach like this.

We might not need to use that in an implementation right now, but it will definitely be a necessary follow-up.